src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -0,0 +1,537 @@
+        ZEROMQ_DEBUG("Listening on log pull socket: %s", listen_log_endpoint.c_str());
+        try {
+            log_pull.bind(listen_log_endpoint);
+        } catch ( zmq::error_t& err ) {
+            zeek::reporter->Error("ZeroMQ: Failed to bind to PULL socket %s: %s", listen_log_endpoint.c_str(),
+                                  err.what());
+            return false;
+        }
+    }
+
+    const auto& log_endpoints = zeek::id::find_val<zeek::VectorVal>("Cluster::Backend::ZeroMQ::connect_log_endpoints");
+    for ( unsigned int i = 0; i < log_endpoints->Size(); i++ )
+        connect_log_endpoints.push_back(log_endpoints->StringValAt(i)->ToStdString());
+
+    for ( const auto& endp : connect_log_endpoints ) {
+        ZEROMQ_DEBUG("Connecting log_push socket with %s", endp.c_str());
+        try {
+            log_push.connect(endp);
+        } catch ( zmq::error_t& err ) {
+            zeek::reporter->Error("ZeroMQ: Failed to connect to PUSH socket %s: %s", endp.c_str(), err.what());
+            return false;
+        }
+    }
+
+    // We may not be connected yet, but there's no logging either unless
+    // we hook into zmq_socket_monitor.
+    //
+    // Check the following if we wanted to add ZeroMQ into the IO loop without
+    // the flare. Maybe?
+    //
+    // https://funcptr.net/2012/09/10/zeromq---edge-triggered-notification/
+    self_thread = std::thread(self_thread_fun, this);
+
+    // After connecting, call ThreadedBackend::DoInit() to register
+    // the IO source with the loop.
+    return ThreadedBackend::DoInit();
+}
+
+bool ZeroMQBackend::SpawnZmqProxyThread() {
+    proxy_thread = std::make_unique<ProxyThread>(listen_xpub_endpoint, listen_xsub_endpoint);
+    return proxy_thread->Start();
+}
+
+bool ZeroMQBackend::DoPublishEvent(const std::string& topic, const std::string& format,
+                                   const cluster::detail::byte_buffer& buf) {
+    // Parts to send for an event publish.
+    std::array<zmq::const_buffer, 4> parts = {
+        zmq::const_buffer(topic.data(), topic.size()),
+        zmq::const_buffer(my_node_id.data(), my_node_id.size()),
+        zmq::const_buffer(format.data(), format.size()),
+        zmq::const_buffer(buf.data(), buf.size()),
+    };
+
+    ZEROMQ_DEBUG("Publishing %zu bytes to %s", buf.size(), topic.c_str());
+
+    for ( size_t i = 0; i < parts.size(); i++ ) {
+        zmq::send_flags flags = zmq::send_flags::none;
+        if ( i < parts.size() - 1 )
+            flags = flags | zmq::send_flags::sndmore;
+
+        // This should never fail, it will instead block
+        // when HWM is reached. I guess we need to see if
+        // and how this can happen :-/
+        main_inproc.send(parts[i], flags);
+    }
+
+    return true;
+}
+
+bool ZeroMQBackend::DoSubscribe(const std::string& topic_prefix) {
+    ZEROMQ_DEBUG("Subscribing to %s", topic_prefix.c_str());
+    try {
+        std::string msg = "\x01" + topic_prefix;
```

Comment:
For XSUB sockets, instead of using setsockopt with `ZMQ_SUBSCRIBE`, one sends a message to the socket with a "\x00" or "\x01" prefix to unsubscribe or subscribe. It's thei XSUB API.